cwrapper/CArachneWrapperCTest.c

```diff
@@ -0,0 +1,149 @@
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "CArachneWrapper.h"
+
+#include <errno.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+
+enum TestReturn { TEST_SKIP, TEST_PASS, TEST_FAIL };
+typedef enum TestReturn TestReturn;
+
+/**
+ * Functions below are used to test thread creation
+ */
+static void*
+funcCreateTest(void* arg) {
+    usleep(1000);
+    *(unsigned*)arg = 0xDEADBEEF;
+    return NULL;
+}
+
+static TestReturn
+createThreadTest(void) {
+    CArachneThreadId threadId;
+    unsigned arg = 0;
+    TestReturn retval = TEST_PASS;
+
+    cArachneInit(NULL, NULL);
+    int ret = cArachneCreateThread(&threadId, funcCreateTest, (void*)&arg);
+    assert(ret ==  0);
+
+    cArachneJoin(&threadId);
+    if (arg != 0xDEADBEEF) {
+        retval = TEST_FAIL;
+    }
+
+    return retval;
+}
+
+/**
+ * Structures are used to collect all test cases
+ */
+typedef TestReturn (*TEST_FUNC)(void);
+struct testcase {
+    const char *description;
+    TEST_FUNC function;
+};
+
+struct testcase testcases[] = {
+    { "createThread", createThreadTest },
+    { NULL, NULL }
+};
+
+/**
+ * Main entrance for all tests
+ */
+int mainTestLoop(int argc, char **argv) {
+    int exitcode = 0;
+    int id = 0, numCases = 0;
+
+    for (numCases = 0; testcases[numCases].description; numCases++) {
+        /* Counting the number of cases*/
+    }
+
+    printf("Testing cases 1..%d\n", numCases);
+
+    for (id = 0; id < numCases; id++) {
+        fflush(stdout);
+
+        TestReturn ret = testcases[id].function();
+        if (ret == TEST_SKIP) {
+            fprintf(stdout, "ok # SKIP %d - %s\n", id + 1, 
+                   testcases[id].description);
+        } else if (ret == TEST_PASS) {
+            fprintf(stdout, "ok %d - %s\n", id + 1, testcases[id].description);
+        } else {
+            fprintf(stdout, "error %d - %s\n", id + 1, 
+                    testcases[id].description);
+            exitcode = 1;
+            break;
+        }
+        fflush(stdout);
+    }
+
+    return exitcode;
+}
+
+/*
+ * Wrapper for main function arguments
+ */
+struct ArgList {
+    int argc;
+    char** argv;
+};
+typedef struct ArgList ArgList;
+
+/*
+ * Wrapper for real main function
+ */
+static void* mainWrapper(void* arg) {
```

Comment:
It looks like the main reason this function exists is to unpack the Arglist. Can we remove this extra layer by just doing the unwrapping in mainTestLoop?

That function could then invoke cArachneShutdown as the last line.

Reply:
Okay, that sounds a clean solution. I will do that.